sql: allow database prefixes for schema names

jayshrivastava opened the pull request:
This PR adds optional database prefixes to schemas for the following ops:
- `CREATE SCHEMA ...`
- `DROP SCHEMA ...`
- `ALTER SCHEMA` (only the source schema can be prefixed to prevent changing the database of a schema using `ALTER SCHEMA thisdb.foo RENAME TO otherdb.bar`)
- `GRANT ... ON SCHEMA ...`
- `REVOKE ... ON SCHEMA ...`
- `SHOW GRANTS ON SCHEMA ...`

Schema names in statements such as `ALTER TYPE t SET SCHEMA` or `ALTER TABLE t SET SCHEMA` etc currently do not take database prefixes (this remains unchanged in this PR) to discourage the creation of cross database relations. Discussion notes for this are [here](https://cockroachlabs.atlassian.net/l/c/CRUiuKYu).

Resolves https://github.com/cockroachdb/cockroach/issues/54295

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55647)
<!-- Reviewable:end -->

jayshrivastava committed:
sql: allow catalog prefix for schema names in CREATE SCHEMA

The CREATE SCHEMA statement was expanded to allow schema
names prefixed with database names. For example, the statement
`CREATE SCHEMA other_db.schema_name` is valid.

Release note (sql change): The CREATE SCHEMA statement was expanded
to allow schema names prefixed with database names.

jayshrivastava committed:
sql: allow catalog prefixes for schema names in DROP SCHEMA

The DROP SCHEMA statement was expanded to allow schema
names prefixed with database names. For example, the statement
`DROP SCHEMA schema_name, other_db.schema_name CASCADE` is valid.

Release note (sql change): The DROP SCHEMA statement was
expanded to allow schema names prefixed with database names.

jayshrivastava committed:
sql: allow database prefix for schema names in ALTER SCHEMA

The ALTER SCHEMA statement was expanded to allow the source
schema to be prefixed with a database name. For example,
`ALTER SCHEMA other_db.schema_name RENAME TO schema_name_1` is valid.

Release note (sql change): The ALTER SCHEMA statement was expanded
to allow the source schema to be prefixed with a database name.

jayshrivastava committed:
sql: allow database prefix for schemas in GRANT, REVOKE, and SHOW GRANTS

The statements GRANT ... ON SCHEMA ...,
REVOKE ... ON SCHEMA ..., and SHOW GRANTS ON SCHEMA ... were all
updated to allow schema names prefixed with database names.

Release note (sql change): The statements GRANT ... ON SCHEMA ...,
REVOKE ... ON SCHEMA ..., and SHOW GRANTS ON SCHEMA ... were all
updated to allow schema names prefixed with database names.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> just the nit

This is really great work!

> Reviewed 7 of 7 files at r1, 9 of 9 files at r2, 6 of 6 files at r3, 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55647#-:-MKeK9zA-RBzsKSYYwSS:b8pfoh1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[pkg/sql/drop_schema.go, line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55647#-MKeI-Ex9ybFUGBM4_d_:-MKeI-Ex9ybFUGBM4_da:b-mgydih) ([raw file](https://github.com/cockroachdb/cockroach/blob/2538b291ce6094307fb0363a648c09737001e6ef/pkg/sql/drop_schema.go#L100)):*
> ```Go
> 	}
> 
> 	if err := d.resolveCollectedObjects(ctx, p, nil); err != nil {
> ```

why is this now nil? maybe add some commentary.

Also, our style guide is that you should comment the parameter when passing nil with `/* db */`


<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r=ajwerner
